lib/gruf/instrumentation/request_logging/hook.rb

```diff
@@ -0,0 +1,116 @@
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+# persons to whom the Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+# OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+#
+require 'socket'
+require_relative 'formatters/base'
+require_relative 'formatters/logstash'
+require_relative 'formatters/plain'
+
+module Gruf
+  module Instrumentation
+    module RequestLogging
+      ##
+      # Represents an error if the formatter does not extend the base formatter
+      #
+      class InvalidFormatterError < StandardError; end
+
+      ##
+      # Handles Rails-style request logging for gruf services.
+      #
+      # This is added by default to gruf servers; if you have `Gruf.use_default_hooks = false`, you can add it back
+      # manually by doing:
+      #
+      #   Gruf::Instrumentation::Registry.add(:request_logging, Gruf::Instrumentation::RequestLogging::Hook)
+      #
+      class Hook < ::Gruf::Instrumentation::Base
+
+        ###
+        # Log the request, sending it to the appropriate formatter
+        #
+        # @return [String]
+        #
+        def call
+          route_key = method_key(call_signature)
+          if success?
+            type = :info
+            status_name = 'GRPC::Ok'
+          else
+            type = :error
+            status_name = response.class.name
+          end
+
+          payload = {}
+          payload[:params] = sanitize(request.to_h) if options.fetch(:log_parameters, false)
+          payload[:message] = success? ? '' : response.message
+          payload[:service] = service_key
+          payload[:method] = payload[:action] = route_key
+          payload[:grpc_status] = status_name
+          payload[:duration] = execution_time.to_f.round(2)
+          payload[:status] = success? ? GRPC::Core::StatusCodes::OK : response.code
+          payload[:thread_id] = Thread.current.object_id
+          payload[:time] = Time.now.to_s
+          payload[:host] = Socket.gethostname
+
+          ::Gruf.logger.send(type, formatter.format(payload))
+        end
+
+        private
+
+        ##
+        # Determine the appropriate formatter for the request logging
+        #
+        # @return [Gruf::Instrumentation::RequestLogging::Formatters::Base]
+        #
+        def formatter
+          fmt = options.fetch(:formatter, :plain)
```

Comment:
Is there a reason why we return a new formatter every time?  Why not memoize this?